iree/hal/metal/metal_buffer.mm

```diff
@@ -0,0 +1,207 @@
+}
+
+// static
+StatusOr<ref_ptr<MetalBuffer>> MetalBuffer::CreateUnretained(
+    MetalDirectAllocator* allocator, MemoryTypeBitfield memory_type,
+    MemoryAccessBitfield allowed_access, BufferUsageBitfield usage, device_size_t allocation_size,
+    device_size_t byte_offset, device_size_t byte_length, id<MTLBuffer> buffer,
+    id<MTLCommandQueue> transfer_queue) {
+  IREE_TRACE_SCOPE0("MetalBuffer::Create");
+  return assign_ref(new MetalBuffer(allocator, memory_type, allowed_access, usage, allocation_size,
+                                    byte_offset, byte_length, buffer, transfer_queue));
+}
+
+MetalBuffer::MetalBuffer(MetalDirectAllocator* allocator, MemoryTypeBitfield memory_type,
+                         MemoryAccessBitfield allowed_access, BufferUsageBitfield usage,
+                         device_size_t allocation_size, device_size_t byte_offset,
+                         device_size_t byte_length, id<MTLBuffer> buffer,
+                         id<MTLCommandQueue> transfer_queue)
+    : Buffer(allocator, memory_type, allowed_access, usage, allocation_size, byte_offset,
+             byte_length),
+      metal_transfer_queue_([transfer_queue retain]),
+      metal_handle_(buffer) {}
+
+MetalBuffer::~MetalBuffer() {
+  IREE_TRACE_SCOPE0("MetalBuffer::dtor");
+  [metal_handle_ release];
+  [metal_transfer_queue_ release];
+}
+
+Status MetalBuffer::FillImpl(device_size_t byte_offset, device_size_t byte_length,
+                             const void* pattern, device_size_t pattern_length) {
+  IREE_ASSIGN_OR_RETURN(auto mapping,
+                        MapMemory<uint8_t>(MemoryAccess::kDiscardWrite, byte_offset, byte_length));
+  void* data_ptr = static_cast<void*>(mapping.mutable_data());
+  switch (pattern_length) {
+    case 1: {
+      uint8_t* data = static_cast<uint8_t*>(data_ptr);
+      uint8_t value_bits = *static_cast<const uint8_t*>(pattern);
+      std::fill_n(data, byte_length, value_bits);
+      break;
+    }
+    case 2: {
+      uint16_t* data = static_cast<uint16_t*>(data_ptr);
+      uint16_t value_bits = *static_cast<const uint16_t*>(pattern);
+      std::fill_n(data, byte_length / sizeof(uint16_t), value_bits);
+      break;
+    }
+    case 4: {
+      uint32_t* data = static_cast<uint32_t*>(data_ptr);
+      uint32_t value_bits = *static_cast<const uint32_t*>(pattern);
+      std::fill_n(data, byte_length / sizeof(uint32_t), value_bits);
+      break;
+    }
+    default:
+      return InvalidArgumentErrorBuilder(IREE_LOC)
+             << "Unsupported scalar data size: " << pattern_length;
+  }
+  if (NeedToAutoSynchronize()) {
+    IREE_RETURN_IF_ERROR(mapping.Flush());
+  }
+  return OkStatus();
+}
+
+Status MetalBuffer::ReadDataImpl(device_size_t source_offset, void* data,
+                                 device_size_t data_length) {
+  IREE_ASSIGN_OR_RETURN(auto mapping,
+                        MapMemory<uint8_t>(MemoryAccess::kRead, source_offset, data_length));
+  if (NeedToAutoSynchronize()) {
```

Comment:
you can do this in the MapMemoryImpl/UnmapMemoryImpl functions instead --- any logic here should be equivalent to map + read/write + unmap, as user code can just do that instead of using the ReadData/WriteData helpers and bypass this requirement. Instead of trying to do this automatically we should just not report that host coherent memory is available when it is not.

Reply:
That requires negotiation using `Allocator::CanAllocate` and `Allocator::MakeCompatible`. I think it's not fully fleshed/used APIs yet? We can consider dropping this once that's ready. But there are also cases the user just want coherent memory. Doing some extra bookkeeping to satisfy that seems okay to me instead of rejecting it.